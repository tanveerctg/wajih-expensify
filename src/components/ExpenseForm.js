import React from 'react';
import moment from 'moment';
import {  SingleDatePicker } from 'react-dates';
import { connect } from 'react-redux';
import uuid from 'uuid'

export class ExpenseForm extends React.Component{

  state={
    description:this.props.expense?this.props.expense.description:'',
    amount:this.props.expense?this.props.expense.amount:'',
    note:this.props.expense?this.props.expense.note:'',
    createdat:this.props.expense?moment(this.props.expense.createdat):moment(),
    focused:false,
    error:false
  }
  onDescriptionChange=(e)=>{
    const description=e.target.value;
    this.setState(()=>({
      description:description
    }))
  }
  onAmountChange=(e)=>{
    const amount=e.target.value;
    this.setState(()=>({
      amount:amount
    }))
  }
  onNoteChange=(e)=>{
    const note=e.target.value;
    this.setState(()=>({
      note
    }))
  }
  onDateChange=(createdat)=>{
    this.setState(()=>({createdat:createdat}))
  };
  onFocusChange=({ focused }) =>{
    this.setState({ focused })
    
  } ;
  handleSubmit=(e)=>{
    if(this.state.description==='' || this.state.amount ===''){
      this.setState(()=>({
        error:true
      }))
    }else{
      this.setState(()=>({
        error:false
      }))
      this.props.onSubmit({
        id:uuid(),
        description:this.state.description,
        note:this.state.note,
        amount:this.state.amount,
        createdat:this.state.createdat.valueOf()
      });
    }
    e.preventDefault();
  }
  render(){

    return(
       <div>         
          <form onSubmit={this.handleSubmit}>
            <input 
              placeholder="Description"
              type="text"
              onChange={this.onDescriptionChange}
              value={this.state.description}
            />
            <input 
              placeholder="Amount"
              type="number"
              onChange={this.onAmountChange}
              value={this.state.amount}
            />
            <textarea 
              placeholder="Add some info (Optional)"
              type="text"
              onChange={this.onNoteChange}
              value={this.state.note}
            />
             <input type="submit" value="Submit"/>
            <SingleDatePicker
              date={this.state.createdat}
              onDateChange={this.onDateChange}
              focused={this.state.focused}
              onFocusChange={this.onFocusChange} 
              showClearDate={this.setState.showClearDate}
              numberOfMonths={1}
              isOutsideRange={()=>false}
            />
           
          </form>
          
          {this.state.error && <p>Please fill up Description and Amount fields</p>}
      </div>
    )
  }
};
export default connect()(ExpenseForm);