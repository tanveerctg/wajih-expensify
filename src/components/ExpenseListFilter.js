import React from 'react';
import { connect } from 'react-redux';
import {set_By_Text} from '../Actions/filters';
import {sortBy_amount} from '../Actions/filters';
import {sortBy_date} from '../Actions/filters';
import {set_By_startDate} from '../Actions/filters';
import {set_By_endDate} from '../Actions/filters';
import { DateRangePicker } from 'react-dates';
import uuid from 'uuid'

class ExpenseListFilter extends React.Component{

  state={
    calendarFocused:null
  }
  onDatesChange=({ startDate, endDate })=>{
    
    this.props.dispatch(set_By_startDate(startDate));
    this.props.dispatch(set_By_endDate(endDate))
  };
  onFocusChange=(calendarFocused)=>{
    this.setState(()=>({
      calendarFocused
    }))
  }
  render(){
   
    return(
      <div className="expenseListFilter">

          <input type="text" onChange={(e)=>{
            this.props.dispatch(set_By_Text(e.target.value));
          }} className="searchExpense" placeholder="Search Expense"/>
  
          <select onChange={(e)=>{
            if(e.target.value==='amount'){
              this.props.dispatch(sortBy_amount(e.target.value));
            }else if(e.target.value==='date'){
              this.props.dispatch(sortBy_date(e.target.value));
            }
          }} className="sortExpense">
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>

          <div className="DateRangePicker">
          <DateRangePicker
            startDate={this.props.filter.startDate} 
            startDateId={uuid()}
            endDate={this.props.filter.endDate} 
            endDateId={uuid()}
            onDatesChange={this.onDatesChange} 
            focusedInput={this.state.calendarFocused} 
            onFocusChange={this.onFocusChange} 
            numberOfMonths={1}
            isOutsideRange={()=>false}
            showClearDates={true}
          />
          </div>
      </div>
    )
  }
}

export default connect(
  (state) => {
    return {
      filter:state.filterReducer
    }
  }
)(ExpenseListFilter);
