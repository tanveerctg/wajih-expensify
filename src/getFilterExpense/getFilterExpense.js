  import moment from 'moment';

  export default (exp,{text='',startDate,endDate,sortBy})=>{

    let hold=exp.filter(expense=>{
    
      const txtMatch =expense.description.toLowerCase().includes(text.toLowerCase());
      const startingDate = startDate ? expense.createdat>= moment(startDate).valueOf() : true;
      const endingDate = endDate ? expense.createdat <= moment(endDate).valueOf() : true;
   
    if(txtMatch && startingDate && endingDate){
      return expense;
    }
    });
  
    if(sortBy==='amount'){
  
      hold=hold.sort((a,b)=>(a.amount<b.amount?1:-1))
    }else if(sortBy==='date'){
      
      hold=hold.sort((a,b)=>(a.createdat<b.createdat?1:-1))
    }
    let sum=0;
    const totalExpense=hold.forEach(item=>{
      sum=parseInt(item.amount)+sum;
    })

    return { hold:hold,
             totalExpense:sum
           };
  
   
  }
