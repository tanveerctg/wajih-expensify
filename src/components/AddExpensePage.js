import React from 'react';
import ExpenseForm from './ExpenseForm';
import {startAddExpense} from '../Actions/expenses';
import { connect } from 'react-redux';
// export const AddExpensePage=(props)=>(
//   <div>
//     <ExpenseForm onSubmit={({id,description,note,createdat,amount})=>{
//      props.dispatch(addExpense({id,description,note,createdat,amount}));

//      props.history.push('/');
//     }}/>
//   </div>
// );
// export default connect()(AddExpensePage);

export class AddExpensePage extends React.Component{
  onSubmit=({id,description,note,createdat,amount})=>{

    this.props.onSubmit({id,description,note,createdat,amount});
    this.props.history.push('/');
  };
  render(){
    return(
      <div>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    )
  }
}
const mapDispatchtoProps=(dispatch)=>{
  return {
    onSubmit:({id,description,note,createdat,amount})=>dispatch(startAddExpense({id,description,note,createdat,amount}))
  }
}
export default connect(undefined,mapDispatchtoProps)(AddExpensePage);
// we can use the following procedure to export function

// export default ()=>(
//   <div>
//     This is from AddExpensePage
//   </div>
// );
