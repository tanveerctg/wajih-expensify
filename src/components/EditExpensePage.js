import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpense} from '../Actions/expenses';
const EditExpensePage=(props)=>(
  <div>

    <ExpenseForm onSubmit={(expense)=>{
       props.dispatch(startEditExpense(props.match.params.id,expense))
       props.history.push('/');
    }} expense={props.expense}/>

  </div>
);
export default connect(
  (state,props) => {
    return {
      expense:state.expenseReducer.find(item=>(item.id===props.match.params.id))
    }
  }
)(EditExpensePage);
