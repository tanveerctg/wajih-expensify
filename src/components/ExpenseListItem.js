import React from 'react';
import { Link } from 'react-router-dom';
import RemoveItem from './RemoveItem';
import { connect } from 'react-redux';
import getFilterExpense from '../getFilterExpense/getFilterExpense';

const ExpenseListItem=({id,description,createdat,amount})=>(
  <div>
    <Link to={`edit/${id}`} ><h3>{description}</h3></Link>
    <p>created at-{createdat}</p>
    <p>amount -{amount}</p>
    <RemoveItem id={id}/>
    <br/>
  </div>
);

export default ExpenseListItem;

