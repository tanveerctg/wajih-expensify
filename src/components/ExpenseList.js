import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getFilterExpense from '../getFilterExpense/getFilterExpense';
export const ExpenseList=(props)=>(
  <div>
    {props.expense.length===0?<h1>No expenses available</h1>: <h1>Expense List</h1>}
    <h2>Total Expense:{props.totalExpense}</h2>
    {
      props.expense.map((exp)=>(<ExpenseListItem key={exp.id} {...exp}/>))   
    }
  </div>
);
export default connect(
  (state) => {
    return {
      expense:getFilterExpense(state.expenseReducer,state.filterReducer).hold,
      totalExpense:getFilterExpense(state.expenseReducer,state.filterReducer).totalExpense
    }
  }
)(ExpenseList);