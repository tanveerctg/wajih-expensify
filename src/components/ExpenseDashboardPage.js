import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import Header from '../components/Header';
const ExpenseDashboardPage=()=>(
  <div>
    <Header/>
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;