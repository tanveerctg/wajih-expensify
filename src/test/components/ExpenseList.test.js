import toJson from 'enzyme-to-json';
import React from 'react';
import { shallow } from 'enzyme';
import {ExpenseList} from  '../../components/ExpenseList';
import setupTests from '../setupTests';
import getFilterExpense from '../getFilterExpense/getFilterExpense.test.js';

test("should render ExpenseList correctly",()=>{
  const wrapper=shallow(<ExpenseList expense={ getFilterExpense }/>);
  expect(toJson(wrapper)).toMatchSnapshot();
})