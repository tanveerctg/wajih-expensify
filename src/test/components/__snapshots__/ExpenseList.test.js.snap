// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList correctly 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={2000}
    createdat={1300}
    description="January"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={4000}
    createdat={900}
    description="May"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={1000}
    createdat={2500}
    description="February"
    id="3"
    key="3"
    note=""
  />
</div>
`;
